Issue #9: Added element and classname configuration options

michaelBenin opened the pull request:
Config properties in constructor:

```
this.markupClassName = config.markupClassName || "___iso-html___";
this.markupElement = config.markupElement || "div";
this.dataClassName = config.dataClassName || "___iso-state___";
this.dataElement = config.dataElement || "div";
```

Also I saw that on render, we create a new Iso, is this intentional? Also added configuration option to the render call as well. 

Examples updated. Manually tested it on date-time. Would like to see more tests, linting, and documentation.

michaelBenin commented:
Ah there's an issue with the date-time. Let me dig into this further.

goatslacker commented on src/iso.js:
```diff
@@ -35,8 +39,8 @@ ${data}
 
-  static render(html, state = {}, meta = {}) {
-    return new Iso().add(html, state, meta).render()
+  static render(config = {markupClassName: '___iso-html___', markupElement: 'div', dataClassName: '___iso-state___', dataElement: 'div'}, html, state = {}, meta = {}) {
```
can we move this to the end so its not a breaking change?

michaelBenin replied:
:+1:

goatslacker commented on src/iso.js:
```diff
@@ -4,7 +4,11 @@ const each = (x, f) => Array.prototype.forEach.call(x, f)
 export default class Iso {
-  constructor() {
+  constructor(config) {
+    this.markupClassName = config.markupClassName || '___iso-html___'
```
i like this :)

michaelBenin commented:
Ok I figured it out, the context of this was broken in the class instance. 

Specifically: https://github.com/goatslacker/iso/pull/15/files#diff-338ebbd84df23821b105676cbab7154bR51

I'm just now moving into using ES6/ES2015/Babel. 

Fixed it with:

```
  static bootstrap(onNode, markupClassName='___iso-html___', dataClassName='___iso-state___') {
```

Is there a better way?

Added those arguements at the end to make it backwards compatible. 

Will update the render method as well. 

Fix should be in later tonight around 8 or 9 EST.

michaelBenin commented:
I'll reopen the pr and squash the commits when it's all cleaned up.

goatslacker commented on src/iso.js:
```diff
@@ -35,17 +39,18 @@ ${data}
   }
 
-  static bootstrap(onNode) {
+  static bootstrap(onNode, markupClassName = '___iso-html___', dataClassName = '___iso-state___') {
```
this should probably be the config object as well

goatslacker replied:
and we can have a function that generates the default options for the config

michaelBenin replied:
Should bootstrap be static? I wasn't able to access the current instance scope here. I originally had it pulling from the config. 

See: https://github.com/michaelBenin/iso/commit/621e7aa6b2b57ae80b020d8f7b8b1d34afe36ae5#diff-4e0c48295f6c143af2dfba2e4bbca24cL76

goatslacker replied:
yeah this is just sugar

otherwise you can create an Iso instance yourself, there's already a bootstrap method on the class.

michaelBenin replied:
To be clear, we do want to remove static from this method.

goatslacker replied:
no they're two different methods, instance bootstrap does the real work where static bootstrap creates an instance and calls bootstrap.

I like the sugar of static bootstrap

goatslacker commented on examples/iso-todomvc/server.js:
```diff
@@ -25,7 +25,7 @@ app.get('/', function (req, res) {
     var markup = React.renderToStaticMarkup(React.createElement(TodoApp))
     res.render('index', {
-      html: Iso.render(markup, data)
+      html: Iso.render({}, markup, data)
```
this needs updating

michaelBenin replied:
Good catch.

goatslacker commented on dist/iso.js:
```diff
@@ -14,32 +14,70 @@ var parse = function (node, x) {
       value: function render(html) {
         var state = arguments[1] === undefined ? {} : arguments[1];
         var meta = arguments[2] === undefined ? {} : arguments[2];
+        var config = arguments[3] === undefined ? { markupClassName: "___iso-html___", markupElement: "div", dataClassName: "___iso-state___", dataElement: "div" } : arguments[3];
```
we need to actually make sure that if someone just provides a markupClass they also get the default for the markupElement and everything else.

michaelBenin replied:
:+1: will do

goatslacker commented:
I might end up switching this from class to prototypes to reduce code size.

Can you update to latest babel on this PR as well?

michaelBenin commented:
The on method should also take the config.

michaelBenin commented:
@goatslacker - this should be good to go. 

Pending we're good, let me know if I should squash the commits.

goatslacker commented:
lgtm
